emission/analysis/modelling/tour_model/featurization.py

```diff
@@ -1,150 +1,137 @@
+        if min_clusters < 2:
+            min_clusters = 2
+        if min_clusters > len(self.points):
+            sys.stderr.write('Maximum number of clusters is the number of data points.\n')
+            min_clusters = len(self.points)-1
+        if max_clusters == None:
+            max_clusters = len(self.points)-1
+        if max_clusters < 2:
+            sys.stderr.write('Must have at least 2 clusters\n')
+            max_clusters = 2
+        if max_clusters >= len(self.points):
+            max_clusters = len(self.points)-1
+        if max_clusters < min_clusters:
+            raise ValueError('Please provide a valid range of cluster sizes')
+        if name != 'kmeans' and name != 'kmedoids':
+            print 'Invalid clustering algorithm name. Defaulting to k-means'
+            name='kmeans'
+        if not self.data:
+            self.sil = None
+            self.clusters = 0
+            self.labels = []
+            return []
+        max = -2
+        num = 0
+        labely = []
+        r = max_clusters - min_clusters+1
+        if name == 'kmedoids':
+            for i in range(r):
+                num_clusters = i + min_clusters
+                print 'testing ' + str(num_clusters) + ' clusters'
+                cl = kmedoids(self.points, num_clusters)
+                self.labels = [0] * len(self.data)
+                cluster = -1
+                for key in cl[2]:
+                    cluster += 1
+                    for j in cl[2][key]:
+                        self.labels[j] = cluster
+                sil = metrics.silhouette_score(numpy.array(self.points), numpy.array(self.labels))
+                if sil > max:
+                    max = sil
+                    num = num_clusters
+                    labely = self.labels
+        elif name == 'kmeans':
+            import warnings
+            for i in range(r):
+                num_clusters = i + min_clusters
+                cl = KMeans(num_clusters, random_state=8)
+                cl.fit(self.points)
+                self.labels = cl.labels_
+                warnings.filterwarnings("ignore")                
+                sil = metrics.silhouette_score(numpy.array(self.points), self.labels)
+                if sil > max:
+                    max = sil
+                    num = num_clusters
+                    labely = self.labels
+        self.sil = max
+        self.clusters = num
+        self.labels = labely
+        self.labels = list(self.labels)
+        return self.labels
+
+    #compute metrics to evaluate clusters
+    def check_clusters(self):
+        if not self.clusters:
+            sys.stderr.write('No clusters to analyze\n')
+            return
+        if not self.labels:
+            print 'Please cluster before analyzing clusters.'
+            return
+        print 'number of clusters is ' + str(self.clusters)
+        print 'silhouette score is ' + str(self.sil) 
+
+    #map the clusters
+    def map_clusters(self):

```

Comment:
map_clusters should be moved out to to a file under emission.analysis.plotting, where it would presumably map the clusters that have been saved to a database? Or can be displayed inline in ipython notebook.